### PR TITLE
Split monolithic module into several collection modules 

### LINKED ISSUE #10670: Split QubesOS Ansible module into multiple modules
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### The problem you're addressing (if any)

Currently, we have a big `qubesos` Ansible module doing various QubesOS related things, such as:

- managing a qube and its properties in an idempotent way: ensure the state of a qube, its properties, create/delete it...
- running some non-idempotent commands by setting the module parameter `command`:
  - to manage a qube state: `create` / `destroy` / `(un)pause` / `shutdown` / `restart` / `remove` 
  - to remove all qube tags: `removetags`
  - to get information about all qubes: `info` / `list_vms` / `get_states`
  - create an Ansible inventory `createinventory`
- listing available devices on the host by setting the module parameter `gather_device_facts`

The module likely manage too many different things and needs to be split into multiple modules to improve its readability and usability, and to facilitate its maintenance.

### The solution you'd like

First, create an Ansible collection for QubesOS plugins (`qubesos.core` ?). 

Create a module dedicated to qube management (`qubesos.core.qube`?). This module must be idempotent and should allow to manage the state of the qube as well as its attributes:
  * `present`: ensure the qube exists, create it if it doesn't
  * `absent`:  ensure the qube doesn't exist, delete if it does
  * `running`, `halted`, `paused`, `restarted`: ensure the qube exists and has the desired state. If one wants for example a qube with `state: running`, the module should create the qube if needed, set the desired properties (label, prefs, devices...), then it should start it.

Some properties (like the template) cannot be set if the qube is running. A special module parameter like `shutdown_if_required` could automatically shutdown the qube before setting those properties.

(Maybe) A module for non-indempotent commands? `qubesos.core.command` for example. It may be useful to have a "kill" command to run in an handler or something else? I don’t see the benefit of keeping commands such as `create`, `pause`, `unpause` but I have doubts about the `restart` command because one may want to restart a Qube but may want to get an error if the qube doesn't exist. Given the description above regarding the `qubes.core.qube` module, `state: restarted` would create the qube and start it: may be counterintuitive...

Write dedicated modules to retrieve facts like host devices, list qubes states, get information about a qube... Those module may be named `qubes.core.host_devices_facts` or `qubesos.core.qube_facts` to [follow Ansible naming convention](https://docs.ansible.com/projects/ansible/latest/dev_guide/developing_modules_best_practices.html#following-ansible-conventions)?

To avoid breaking changes as much as possible, the legacy `qubesos` module should remain in place and should try to translate old parameters to calls to the new modules with appropriates parameters and return unchanged data structures (at least for the facts). It should also display a depreciation warning to encourage users to switch to the new modules.

### The value to a user and who that user might be

More user-friendly modules reducing the risk of errors. Adding features is easier, for example, if we want to manage QubesOS label, we can simply write a dedicated module doing it without carrying about other modules.

More readable code, making maintenance easier. 

### Completion criteria checklist

(This section is for developer use only. Please do not modify it.)

### PR DESCRIPTION
**Pull request by Guiiix**
Fixes https://github.com/QubesOS/qubes-issues/issues/10670

This PR brings some architectural changes to this repository.

## 1. Introduction of collections

Ansible collections allow to distribute a set of plugins, roles and modules into a self-contained, reusable and versioned package.

The `qubesos` module allow to do a lot of different things, many of which are not directly related. Splitting it into 3 modules helps for maintenance and for usage in playbooks.
2 collections are introduced:
- `qubesos.core` for everything related to QubesOS management
- `qubesos.security` for Ansible-related security features (basically, the `qubes-proxy` and the `qubesos_strategy_guard`)

The `qubesos` module has been splitted into 3 different modules:
- `qubesos.core.qube`: to manage the qubes
- `qubesos.core.command`: non indempotent commands (required to keep compatibility with the old module)
- `qubesos.core.host_devices_facts`: gather facts about available devices on the host

Other plugins (connection, strategy and callback) have been moved into a collection but may still be called using their old name.

| Plugin type | Old name | New name |
| ---------------- | --------------- | --------------- |
| Connection | `qubes`   | `qubesos.core.qubes` |
| Strategy | `qubes_proxy` | `qubesos.security.qubes_proxy` |
| Callback | `qubes_strategy_guard` | `qubesos.security.qubes_strategy_guard` |

## 2. Legacy module

To prevent breaking changes, the `qubesos` module is still present but is considered deprecated and may be removed in a future release.

The module takes the same options and will try to translate to calls to the new 
modules with the appropriate options.

Option `wait` has no more effect. The module will always wait for the actions (qube start, stop...)
to finish before starting a new task to prevent unexpected behaviors.

## 3. A word about `module_utils`

The files in the `modules` directory contains only the module documentation while the entire module code is stored in the directory `module_utils`. The reason of that split is the legacy module `qubesos`. As described above, this module is responsible of taking the same input as the previous module and based on the input, it will call the new modules.

However, [Ansiballz framework](https://docs.ansible.com/projects/ansible/latest/dev_guide/developing_program_flow_modules.html#ansiballz-framework) doesn't include other modules in the generated archive making imports fail. Putting modules logic into `module_utils` make the code available to other modules, and so available to `qubesos` module.

## 4. Documentation

The documentation is now automatically generated based on [modules docstrings](https://docs.ansible.com/projects/ansible/latest/dev_guide/developing_modules_documenting.html) and is available in the [project Gitlab Pages](https://qubesos.gitlab.io/qubes-ansible).

Focusing on documenting modules using the official way allow to generate the documentation with the common Ansible format while also making it available using `ansible-docs` command.

### CODE REVIEW

**Comment by Guiiix** [comment]
PipelineRetry

**Comment by marmarek** [comment]
> * gpu_passthrough: [Failed](https://openqa.qubes-os.org/tests/173356#step/gpu_passthrough/31) (test died)
>       `# Test died: command 'ansible-playbook -i inventory -e pci_devices=...`

Old playbook stopped working. It's here: https://github.com/QubesOS/openqa-tests-qubesos/tree/main/extra-files/ansible

File: ansible_collections/qubesos/core/plugins/module_utils/qubes_module_qube.py [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+        # Check VM existence for the following properties
+
+        for property in [
+            "audiovm",
+            "default_dispvm",
+            "management_dispvm",
+            "netvm",
+        ]:
+            value = self.wants.properties.get(property)
+            if value in (None, "dom0", "*default*"):
```
Comment (marmarek) [inline ansible_collections/qubesos/core/plugins/module_utils/qubes_module_qube.py]: It used to accept also empty string as None (existing playbooks rely on that).

File: plugins/modules/qubesos.py [hunk 2]
```diff
@@ -302,727 +351,289 @@ def create_inventory(result):
...
+    # Check netvm exist
+    # Check volume properties
+    # This is only verification: should be conserved as format is different from
+    # qubesos.core.qube
     if properties:
         for key, val in properties.items():
             if key not in PROPS:
-                return VIRT_FAILED, {"Invalid property": key}
+                raise ValidationFailure({"Invalid property": key})
             if type(val) != PROPS[key]:
```
Comment (marmarek) [inline plugins/modules/qubesos.py]: This is unfortunate given the new module doesn't accept empty string as None anymore, yet this line rejects real None...

File: ansible_collections/qubesos/core/plugins/module_utils/qubes_module_qube.py (hunk 1 above)
Comment (marmarek) [reply ansible_collections/qubesos/core/plugins/module_utils/qubes_module_qube.py]: This probably also wants a test for setting any of the VM-type properties to None.

File: ansible_collections/qubesos/core/plugins/module_utils/qubes_module_qube.py [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+                continue
+
+            try:
+                vm = self.helper.get_vm(value)
```
Comment (marmarek) [inline ansible_collections/qubesos/core/plugins/module_utils/qubes_module_qube.py:499]: This validation is too early. This prevents creating a (disposable template) qube that has itself as its `default_dispvm`. It used to be possible.

**Review by marmarek — CHANGES_REQUESTED** [review]

**Comment by Guiiix** [comment]
Thanks for the review @marmarek.

Should be good for the openqa playbook. Ansible doesn't like Enum in returned data...

Now we can set an empty string or None in a qube property and qube existence is skipped when specifying the managed qube in a property.

Some tests added too.

**Review by marmarek — APPROVED** [review]
